apps/ejabberd/src/http_upload/mod_http_upload.erl

```diff
@@ -0,0 +1,128 @@
+%%==============================================================================
+%% Copyright 2016 Erlang Solutions Ltd.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%==============================================================================
+
+-module(mod_http_upload).
+-author('konrad.zemek@gmail.com').
+-behaviour(gen_mod).
+-xep([{xep, 363}, {version, "0.2.4"}]).
+
+-include("jlib.hrl").
+-include("ejabberd.hrl").
+
+-export([start/2, stop/1, iq_handler/3]).
+
+%%--------------------------------------------------------------------
+%% Callbacks
+%%--------------------------------------------------------------------
+
+-callback create_slot(UTCDateTime :: calendar:datetime(), UUID :: binary(),
+                      Filename :: unicode:unicode_binary(), ContentType :: binary(),
+                      Size :: pos_integer(), Opts :: proplists:proplist()) ->
+    {PUTURL :: binary(), GETURL :: binary()}.
+
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+
+-spec start(Host :: ejabberd:server(), Opts :: list()) -> any().
+start(Host, Opts) ->
+    IQDisc = gen_mod:get_opt(iqdisc, Opts, one_queue),
+    mod_disco:register_feature(Host, ?NS_HTTP_UPLOAD),
+    gen_iq_handler:add_iq_handler(ejabberd_local, Host, ?NS_HTTP_UPLOAD, ?MODULE,
+                                  iq_handler, IQDisc).
+
+
+-spec stop(Host :: ejabberd:server()) -> any().
+stop(Host) ->
+    gen_iq_handler:remove_iq_handler(ejabberd_local, Host, ?NS_HTTP_UPLOAD),
+    mod_disco:unregister_feature(Host, ?NS_HTTP_UPLOAD).
+
+
+-spec iq_handler(From :: ejabberd:jid(), To :: ejabberd:jid(), IQ :: ejabberd:iq()) ->
+                        ejabberd:iq() | ignore.
+iq_handler(_From, _To, IQ = #iq{type = set, sub_el = SubEl}) ->
+    IQ#iq{type = error, sub_el = [SubEl, ?ERR_NOT_ALLOWED]};
+iq_handler(_From, _To, IQ = #iq{type = get, sub_el = Request}) ->
+    case parse_request(Request) of
+        {Filename, Size, ContentType} ->
+            UTCDateTime = calendar:universal_time(),
+            Token = generate_token(),
+            BackendModule = backend_module(),
```

Comment:
I would consider using `gen_mod:start_backend_module/2` to be more consistent with other modules with swappable backends. This will also automatically maintain a metric on the call to `create_slot`. :)